models/models_methods.go

```diff
@@ -462,38 +464,105 @@ func validateStep(stepID string, step stepmanModels.StepModel) ([]string, error)
 
 func validatePipelines(config *BitriseDataModel) ([]string, error) {
 	pipelineWarnings := make([]string, 0)
-	for ID, pipeline := range config.Pipelines {
-		idWarning, err := validateID(ID, "pipeline")
+	for pipelineID, pipeline := range config.Pipelines {
+		idWarning, err := validateID(pipelineID, "pipeline")
 		if idWarning != "" {
 			pipelineWarnings = append(pipelineWarnings, idWarning)
 		}
 		if err != nil {
 			return pipelineWarnings, err
 		}
 
-		if len(pipeline.Stages) == 0 {
-			return pipelineWarnings, fmt.Errorf("pipeline (%s) should have at least 1 stage", ID)
+		hasStages := len(pipeline.Stages) > 0
+		hasWorkflows := len(pipeline.Workflows) > 0
+
+		if hasStages && hasWorkflows {
+			return pipelineWarnings, fmt.Errorf("pipeline (%s) has both stages and workflows", pipelineID)
+		} else if !hasStages && !hasWorkflows {
+			return pipelineWarnings, fmt.Errorf("pipeline (%s) should have at least 1 stage or workflow", pipelineID)
 		}
 
-		for _, pipelineStage := range pipeline.Stages {
-			pipelineStageID, err := getStageID(pipelineStage)
-			if err != nil {
-				return pipelineWarnings, err
+		if hasStages {
+			return pipelineWarnings, validateStagedPipeline(pipelineID, &pipeline, config)
+		}
+
+		//TODO: Why is this always true?
+		if hasWorkflows {
+			return pipelineWarnings, validateDAGPipeline(pipelineID, &pipeline, config)
+		}
+
+	}
+
+	return pipelineWarnings, nil
+}
+
+func validateStagedPipeline(pipelineID string, pipeline *PipelineModel, config *BitriseDataModel) error {
+	for _, pipelineStage := range pipeline.Stages {
+		pipelineStageID, err := getStageID(pipelineStage)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := config.Stages[pipelineStageID]; !ok {
+			return fmt.Errorf("stage (%s) defined in pipeline (%s), but does not exist", pipelineStageID, pipelineID)
+		}
+	}
+
+	return nil
+}
+
+func validateDAGPipeline(pipelineID string, pipeline *PipelineModel, config *BitriseDataModel) error {
+	for pipelineWorkflowID, pipelineWorkflow := range pipeline.Workflows {
+		if isUtilityWorkflow(pipelineWorkflowID) {
+			return fmt.Errorf("workflow (%s) defined in pipeline (%s), is a utility workflow", pipelineWorkflowID, pipelineID)
+		}
+
+		if _, ok := config.Workflows[pipelineWorkflowID]; !ok {
+			return fmt.Errorf("workflow (%s) defined in pipeline (%s), but does not exist", pipelineWorkflowID, pipelineID)
+		}
+
+		uniqueItems := make(map[string]struct{})
```

Comment:
I always use the most memory efficient one because all of the blog post used that one. But I have to admit I did not think it through what it means to use a boolean value instead. I quite like that and will try it out.